apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/backend/ir/BackendIrBuilder.kt

```diff
@@ -452,144 +410,161 @@ internal class BackendIrBuilder constructor(
+  private fun List<BackendIr.Fragment>.buildFragmentImplementation(
+      parentName: String,
+      parentSelectionSet: List<BackendIr.Field>,
+      possibleTypes: Set<IntrospectionSchema.TypeRef>,
+      selectionKey: SelectionKey,
+  ): BackendIr.Fragment {
+    val fragmentName = this.distinctBy { fragment -> fragment.name }
+        .joinToString(separator = "", postfix = parentName.capitalize()) { fragment -> fragment.name.capitalize() }
+
+    val selectionSet = this.fold(parentSelectionSet) { acc, fragment ->
+      acc.mergeFields(
+          fragment.fields.addFieldSelectionKey(selectionKey + fragmentName)
+      )
+    }.map { field ->
+      field.buildFragmentImplementations(
+          selectionKey = selectionKey + fragmentName + field.responseName,
+      )
     }
+
+    val selectionsKeys = this.fold(emptySet<SelectionKey>()) { acc, fragment ->
+      acc.plus(fragment.selectionKeys)
+    }.plus(selectionKey)
+
+    return BackendIr.Fragment(
+        name = fragmentName,
+        fields = selectionSet,
+        possibleTypes = possibleTypes.toSet(),
+        selectionKeys = selectionsKeys,
+        description = null,
+        kind = BackendIr.Fragment.Kind.Implementation,
+    )
   }
 
   private fun buildGenericFragments(
       inlineFragments: List<FrontendIr.Selection.InlineFragment>,
       namedFragments: List<FrontendIr.Selection.FragmentSpread>,
-      selectionKey: SelectionKey,
-      selectionSet: List<BackendIr.Field>,
-      generateFragmentImplementations: Boolean,
+      parentSelectionKey: SelectionKey,
+      parentSelectionSet: List<BackendIr.Field>,
+      parentNamedFragmentSelectionKey: SelectionKey? = null,
   ): List<GenericFragment> {
-    // build generic fragments from inline fragments
     val genericInlineFragments = inlineFragments
         .map { inlineFragment ->
-          inlineFragment.buildGenericFragment(
-              parentSelectionKey = selectionKey,
-              parentSelectionSet = selectionSet,
-              parentNamedFragmentSelectionKeys = emptySet(),
-              generateFragmentImplementations = generateFragmentImplementations
+          val fragmentName = inlineFragment.fragmentDefinition.typeCondition.name.capitalize()
+          buildGenericFragment(
+              fragmentName = fragmentName,
+              fragmentTypeCondition = inlineFragment.fragmentDefinition.typeCondition.name,
+              fragmentSelectionSet = inlineFragment.fragmentDefinition.selections,
+              fragmentDescription = inlineFragment.fragmentDefinition.typeCondition.description ?: "",
+              fragmentCondition = inlineFragment.condition.toBackendIrCondition(),
+              namedFragmentSelectionKey = parentNamedFragmentSelectionKey?.plus(fragmentName),
+              parentSelectionKey = parentSelectionKey,
+              parentSelectionSet = parentSelectionSet.addFieldSelectionKey(
+                  parentNamedFragmentSelectionKey?.plus(fragmentName)
+              ),
```

Comment:
I hate to be the bearer of bad news but we'll need to recurse nested inline fragments as well. One exemple is there: https://github.com/apollographql/apollo-android/commit/2c2b39ad670f31a51e072701c51b035d28f64c9d

Reply:
should be fixed with this commit https://github.com/apollographql/apollo-android/pull/2806/commits/a68fb53966b6626e653d061a3ae3eca7dd524d62#diff-7cee39198b0604492035a9fe87f2c7c4395fe30c6bad0c7a312849e4f73a2da4R171